### PR TITLE
Feature/tao 4905 persistent audio volume

### PR DESCRIPTION
**Pull request by tikhanovichA**

### CODE REVIEW

File: views/js/ui/mediaplayer.js [hunk 1]
```diff
@@ -1874,6 +1884,40 @@ define([
...
 
         /**
+         * Update volume in DBIndex store
+         * @param volume
+         * @private
+         */
+        _storeVolume: function _storeVolume(volume) {
+            return store('mediaVolume')
+                .then(function(volumeStore) {
+                    return volumeStore.clear().then(function() {
```
Comment (krampstudio) [inline views/js/ui/mediaplayer.js]: You don't need to `clear` the store if you do a `setItem` just after

File: views/js/ui/mediaplayer.js [hunk 2]
```diff
@@ -1874,6 +1884,40 @@ define([
...
+         * Get volume from DBIndex store
+         * @private
+         */
+        _getStoredVolume: function _getStoredVolume() {
+            var self = this;
+            return store('mediaVolume')
+                .then(function(volumeStore) {
+                    return volumeStore.getItem('volume')
+                        .then(function(volume) {
+                            if (typeof(volume) !== 'undefined' && jQuery.type !== 'object') {
```
Comment (krampstudio) [inline views/js/ui/mediaplayer.js]: `jQuery.type` ?

File: manifest.php [hunk 3]
```diff
@@ -36,7 +36,7 @@
     'label' => 'Tao base',
     'description' => 'TAO meta-extension',
     'license' => 'GPL-2.0',
-    'version' => '12.8.2',
+    'version' => '12.9.2',
```
Comment (krampstudio) [inline manifest.php]: increasing the feature number, always resets the hotfix numver to 0 in semver. You should increase 
`12.8.2` to `12.9.0`

File: views/js/ui/mediaplayer.js [hunk 4]
```diff
@@ -1874,6 +1884,40 @@ define([
         },
 
         /**
+         * Update volume in DBIndex store
+         * @param volume
```
Comment (krampstudio) [inline views/js/ui/mediaplayer.js]: If we expect `volume` to be a number, the documentation should be : 

```
/**
 * Store the volume 
 * @param {Number} volume
 */
```

File: views/js/ui/mediaplayer.js (hunk 2 above)
Comment (krampstudio) [inline views/js/ui/mediaplayer.js]: I'd instead check the type  and apply the rule : 

```js
if(_.isNumber(volume)){
    self.volume = Math.max(_volumeMin, Math.min(_volumeMax, parseFloat(volume)));
}
```

**Comment by tikhanovichA** [comment]
Fixed issues and updated version.

File: views/js/ui/mediaplayer.js [hunk 5]
```diff
@@ -1698,10 +1702,16 @@ define([
                 self.seek(value, true);
             });
 
+            $(document).on('updateVolume', function(event, value) {
+                self.setVolume(value);
+            });
+
             this.$volume.on('change' + _ns, function(event, value) {
                 self.unmute();
+                $(document).trigger('updateVolume', value);
```
Comment (eacraft13) [inline views/js/ui/mediaplayer.js]: Namespace the event: `'updateVolume' + _ns`

File: views/js/ui/mediaplayer.js [hunk 6]
```diff
@@ -1698,10 +1702,16 @@ define([
                 self.seek(value, true);
             });
 
+            $(document).on('updateVolume', function(event, value) {
```
Comment (eacraft13) [inline views/js/ui/mediaplayer.js]: Unbind this event on 'destroy'. There is an `_unbindEvents` method already.

File: views/js/ui/mediaplayer.js [hunk 7]
```diff
@@ -1596,6 +1597,9 @@ define([
             this.autoStart = this.config.autoStart;
             this.autoStartAt = this.config.autoStartAt;
             this.startMuted = this.config.startMuted;
+
+            this._getStoredVolume();
```
Comment (eacraft13) [inline views/js/ui/mediaplayer.js]: Could this be moved to `init()`? `_reset()` is called on `destroy`, which doesn't need to retrieve the stored volume.

File: views/js/ui/mediaplayer.js [hunk 8]
```diff
@@ -1874,6 +1884,35 @@ define([
...
+                .then(function(volumeStore){
+                    volumeStore.setItem('volume', volume);
+                });
+        },
+
+        /**
+         * Get volume from DBIndex store
+         * @private
+         */
+        _getStoredVolume: function _getStoredVolume() {
```
Comment (eacraft13) [inline views/js/ui/mediaplayer.js]: This isn't a requirement, but I'd like to see the following change...

The name `_getStoredVolume` is misleading. Perhaps `_updateVolumeFromStore` is more appropriate.

File: views/js/ui/mediaplayer.js [hunk 9]
```diff
@@ -1874,6 +1884,35 @@ define([
...
+            var self = this;
+            return store('mediaVolume')
+                .then(function(volumeStore) {
+                    return volumeStore.getItem('volume')
+                        .then(function(volume) {
```
Comment (eacraft13) [inline views/js/ui/mediaplayer.js]: I believe these can be chained better to not have so much indentation. If `getItem()` returns a promise, you can simply return that and still use a `then`. 

Better explained through example...

```
store('mediaVolume')
.then(function (volumeStore) {
    return volumeStore.getItem('volume');
})
.then(function (volume) {
    ...
});
```

... I believe this should work.

**Comment by tikhanovichA** [comment]
Fixed

**Pull request merged**